File: backend-java/src/main/java/com/asmolabs/zanshin/container/services/ContainerService.java
package com.asmolabs.zanshin.container.services;

import com.asmolabs.zanshin.container.dto.CreateContainerDto;
import com.asmolabs.zanshin.container.entities.Container;
import com.asmolabs.zanshin.container.repositories.ContainerRepository;
import com.asmolabs.zanshin.repository.entities.Scan;
import com.asmolabs.zanshin.repository.repositories.ScanRepository;
import com.asmolabs.zanshin.repository.services.ScanProcessor;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
@Slf4j
public class ContainerService {

    private final ContainerRepository containerRepository;
    private final ScanRepository scanRepository;
    private final ScanProcessor scanProcessor;

    @Transactional
    public Container create(CreateContainerDto dto) {
        String imageName = dto.imageName();
        String tag = dto.tag();

        if (imageName != null && imageName.contains(":")) {
            String[] parts = imageName.split(":");
            imageName = parts[0];
            if (parts.length > 1) {
                tag = parts[1];
            }
        }

        Container container = Container.builder()
                .registry(dto.registry())
                .imageName(imageName)
                .tag(tag != null ? tag : "latest")
                .scanIntervalMinutes(dto.scanIntervalMinutes())
                .scanCron(dto.scanCron())
                .build();

        return containerRepository.save(container);
    }

    public List<Container> findAll() {
        return containerRepository.findAll();
    }

    public Optional<Container> findOne(Long id) {
        return containerRepository.findById(id);
    }

    @Transactional
    public void remove(Long id) {
        containerRepository.deleteById(id);
    }

    @Transactional
    public Scan triggerRescan(Long id) {
        Container container = containerRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Container not found"));

        Scan scan = Scan.builder()
                .container(container)
                .status("pending")
                .branch(container.getTag() != null ? container.getTag() : "latest")
                .subPath("")
                .build();
        
        scan = scanRepository.save(scan);

        // Notify processor
        scanProcessor.processScan(scan.getId(), null, scan.getBranch(), "", null);
        
        return scan;
    }
}

File: zanshin/services/container_service.py
from typing import List
from zanshin.models.container import Container
from zanshin.models.scan import Scan
from zanshin.repositories.container_repository import ContainerRepository
from zanshin.repositories.scan_repository import ScanRepository
from zanshin.services.scan_processor import ScanProcessor
from zanshin.clock import utcnow
from zanshin.services.repository_service import executor

class ContainerService:
    def __init__(self, container_repository: ContainerRepository, scan_repository: ScanRepository, scan_processor: ScanProcessor):
        self.container_repository = container_repository
        self.scan_repository = scan_repository
        self.scan_processor = scan_processor

    def find_all(self) -> List[Container]:
        return self.container_repository.find_all()

    def find_by_id(self, container_id: int) -> Container:
        return self.container_repository.find_by_id(container_id)

    def save(self, container: Container) -> Container:
        return self.container_repository.save(container)

    def delete_by_id(self, container_id: int):
        return self.container_repository.delete_by_id(container_id)

    def trigger_scan(self, container_id: int) -> Scan:
        container = self.container_repository.find_by_id(container_id)
        if not container:
            raise RuntimeError("Container not found")

        scan = Scan(
            container_id=container.id,
            branch=container.tag if container.tag else "latest",
            sub_path="",
            status="pending",
            findings_count=0,
            created_at=utcnow()
        )
        scan = self.scan_repository.save(scan)

        # Straight to the shared pool — see the note in RepositoryService about
        # why the deprecated `get_event_loop()` detour is gone.
        executor.submit(
            self.scan_processor.process_scan,
            scan.id,
            None,
            scan.branch,
            "",
            None,
        )
        return scan
